### PR TITLE
Libvirt metadata vm_id usage

### PR DESCRIPTION
**Pull request by aharivel**
This should answer the Enhancement #1625  

By default the name chosen to define the Virtual Machine ID is the one found in /proc/\<pid\>/cgroup/.
Many IaaS platform like Openstack use metadata in libvirt to customize among other things the instance's ID.
Allow the user to use the metadata in libvirt to set the name of the vm metrics and gives a better user experience.
    
    - Set "LIBVIRT_METADATA_URI" with the uri of the XML namespace identifier.
    - Set "LIBVIRT_METADATA_TOKEN" to choose the string used for VM ID. By default, the token "name" is used.
 
If no URI is set or if any errors occurs, the default method to set the name is kept.

Before: 
```
# HELP kepler_vm_bpf_block_irq_total Aggregated value in bpf_block_irq value from bpf
# TYPE kepler_vm_bpf_block_irq_total counter
kepler_vm_bpf_block_irq_total{source="bpf",vm_id="machine-qemu-1-fedora-40"} 0
kepler_vm_bpf_block_irq_total{source="bpf",vm_id="machine-qemu-2-alpinelinux3.19"} 0
# HELP kepler_vm_bpf_cpu_time_ms_total Aggregated value in bpf_cpu_time_ms value from bpf
# TYPE kepler_vm_bpf_cpu_time_ms_total counter
kepler_vm_bpf_cpu_time_ms_total{source="bpf",vm_id="machine-qemu-1-fedora-40"} 113
kepler_vm_bpf_cpu_time_ms_total{source="bpf",vm_id="machine-qemu-2-alpinelinux3.19"} 25
[...]

```
After
```
# HELP kepler_vm_bpf_block_irq_total Aggregated value in bpf_block_irq value from bpf
# TYPE kepler_vm_bpf_block_irq_total counter
kepler_vm_bpf_block_irq_total{source="bpf",vm_id="instance-alpine"} 0
kepler_vm_bpf_block_irq_total{source="bpf",vm_id="machine-qemu-1-fedora-40"} 0
# HELP kepler_vm_bpf_cpu_time_ms_total Aggregated value in bpf_cpu_time_ms value from bpf
# TYPE kepler_vm_bpf_cpu_time_ms_total counter
kepler_vm_bpf_cpu_time_ms_total{source="bpf",vm_id="instance-alpine"} 0
kepler_vm_bpf_cpu_time_ms_total{source="bpf",vm_id="machine-qemu-1-fedora-40"} 577
[...]
```

This is obtains with the "instance-alpine" libvirt XML :
```
<domain type='kvm' id='2'>
  <name>alpinelinux3.19</name>
  <uuid>45885a2b-6e3a-4b8a-bfbc-178942gf803</uuid>
  <metadata>
    <libosinfo:libosinfo xmlns:libosinfo="http://libosinfo.org/xmlns/libvirt/domain/1.0">
      <libosinfo:os id="http://alpinelinux.org/alpinelinux/3.19"/>
    </libosinfo:libosinfo>
    <nova:instance xmlns:nova="http://openstack.org/xmlns/libvirt/nova/1.1">
      <nova:package version="27.3.1-18.0.202445915514.dd48561540.el9osttrunk"/>
      <nova:name>instance-alpine</nova:name>
      <nova:creationTime>2024-07-11 16:41:29</nova:creationTime>
      <nova:flavor name="alpine_base_flavor">
        <nova:memory>4096</nova:memory>
        <nova:disk>20</nova:disk>
        <nova:swap>0</nova:swap>
        <nova:ephemeral>0</nova:ephemeral>
        <nova:vcpus>4</nova:vcpus>
      </nova:flavor>
      <nova:owner>
        <nova:user uuid="8abfae245945675313521fa574e94b">admin</nova:user>
        <nova:project uuid="3456f07e078789546318104d29c2868f59">admin</nova:project>
      </nova:owner>
      <nova:root type="image" uuid="56aa78956h45-2856-49e7-ajdk-b748946"/>
      <nova:ports>
        <nova:port uuid="78895418-727d-4dac-86c8-382345456">
          <nova:ip type="fixed" address="10.10.10.10" ipVersion="4"/>
        </nova:port>
        <nova:port uuid="43d7897895e6-ca52-44e5-b2878-7864153">
          <nova:ip type="fixed" address="10.10.20.10" ipVersion="4"/>
        </nova:port>
      </nova:ports>
    </nova:instance>
  </metadata>
  <memory unit='KiB'>2097152</memory>
[...]
```

### CODE REVIEW

**Review by dave-tucker — CHANGES_REQUESTED** [review]
The libvirt dependency and performance of this code are likely to be deal breakers for me.

I hear what you say in https://github.com/sustainable-computing-io/kepler/issues/1625, but maybe there's another way we an achieve what you're looking for?

e.g what if instead of reading the cgroup we were to instead look at:

```
$ xargs --null -n1 -a /proc/3905059/cmdline | grep uuid -n1
20-4,sockets=4,cores=1,threads=1
21:-uuid
22-8fc5d2df-80d0-4971-a198-d2fc2602f5f7
```

Which gets the dom uuid (avoiding having to list all + search).

Another alternative might be to add all visited doms to the cache (assuming we're able to get mapping to a pid/tgid).

Using https://github.com/digitalocean/go-libvirt would get around the Cgo concerns.

**Comment by aharivel** [comment]
Hi @dave-tucker 
First, thanks for your feedback.

So if I resume what we have done so far to detect which processes are a VM:
- 10 month ago I proposed to read "/run/libvirt/qemu/<name>.pid" to match the pid<-> name
- 4 month later @marceloamaral changed it to read "/proc/<pid>/cgroup" to unify with the containers filtering
- now you propose to filter /proc/<pid>/cmdline to get uuid of the VM.

Some questions come to my mind:
- How do you know that that <pid> is a VM before running the command you proposed ? Are we going to run it for any processes (!!!) or are we first read the "/proc/<pid>/cgroup" then get the uuid of the VM by reading the cmdline ?
- We have a map[uint64]string that gives the pair "pid<->name" so we would change the name by uuid ? i.e "pid<->uuid"
- I can check the digitalocean libvirt library if that suits the performance better. But in the end, if we don't do anything with "libvirt" (not reading libvirt folders, nor using libvirt librairies) we should change the name of this folder, because it's really misleading.

File: go.mod
Comment (aharivel) [reply go.mod:11]: Only My habit to use this library. I'll try encoding/xml then !

File: go.mod
Comment (aharivel) [reply go.mod:30]: If we can go with https://github.com/digitalocean/go-libvirt then we might be able to achieve the same result. I'm going to check this.

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:128]: Kepler is running as a privileged user to be able to use eBPF so this is no problem

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:135]: I'll check this point with the uuid and digitalocean lib

File: go.mod
Comment (aharivel) [reply go.mod:30]: Ok so this small example: 
```
package main

import (
	"fmt"
	"log"
	"net/url"

	"github.com/digitalocean/go-libvirt"
)

func main() {
	uuid_t := libvirt.UUID{0x01, 0x88, 0x5a, 0x2b, 0x6e, 0x3a, 0x4b, 0x8a, 0xbf, 0xbc, 0x13, 0x72, 0x05, 0xc3, 0xc8, 0x03}

	uri, _ := url.Parse(string(libvirt.QEMUSystem))
	l, err := libvirt.ConnectToURI(uri)
	if err != nil {
		log.Fatalf("failed to connect: %v", err)
	}

	d, errr := l.DomainLookupByUUID(uuid_t)
	if errr != nil {
		log.Fatalf("Domainlookup fail %v", errr)
	}

	fmt.Println("ID\tName\t\tUUID")
	fmt.Printf("--------------------------------------------------------\n")
	fmt.Printf("%d\t%s\t%x\n", d.ID, d.Name, d.UUID)
}
```
is returning me :
$ go run main.go 
ID	Name		UUID
--------------------------------------------------------
1	alpinelinux3.19	01885a2b6e3a4b8abfbc137205c3c803

Comment (aharivel) [reply go.mod:30]: So we could straight after do something like this: 
```
uri := libvirt.optstring{"http://openstack.org/xmlns/libvirt/nova/1.1"}                                                              
str, _ := l.domaingetmetadata(d, 0x2, uri, 0)  
```
and we received the metadata, then we filter with the right token.

Comment (aharivel) [reply go.mod:30]: So in the end, we trade a C-go calls to RPC calls. I'm not that of an expert in this kind of optimization so I'll go with what you prefer.

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:128]: PS: all my smalls tests (digitalocean libvirt) above are done without any modification. So I guess that it works without any privilege.

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:135]: See my comments above that remove this.

File: go.mod
Comment (aharivel) [reply go.mod:11]: @dave-tucker The issue here is that we don't know in advance the struct of the XML. So no possible way to use Unmarshall because it maps XML element to a struct. 
Etree lets you create a document and then FindElement into it.
If you have a better idea, please tell me, I'm making a second version now.

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:123]:
```suggestion
		return "", fmt.Errorf("the vmid does not match machine-qemu-number-name` : %v", vmid)
```

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:122]: won't there be 4 parts?

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:126]:
```suggestion
	} 
	
	domainName = strings.Join(parts[3:], "-")

```

**Comment by sthaha** [comment]
I am marking this with `do-not-merge` since this change needs to be tested manually.

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:81]:
```suggestion
	if config.GetLibvirtMetadataURI() != "" {
		// NOTE: use VM Metadata instead 	
		if metaID, err := getMetadataVMID(vmID); metaID != "" && err == nil {
			vmID = metaID
		} 
	}
	addVMIDToCache(pid, vmID)
```

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:118]: can we move to close to the usage?

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:144]: Doesn't this re-iterate all domains (say 100)  even if we the vm was found in the first iteration in the loop below? 

I think we should simplify the loop below as 

```
for _, dom := range domains {
  name, err := parseDomainForVmID(dom)
  dom.Free()

  if err == nil {
    return name
  }
}

return "", fmt.Errorf("...")
```

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:118]: Thanks for your review. I will include all of your feedback  in the next iteration. 
However I don't understand this question. Could you please elaborate ?

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:144]: On my next iteration using the libvirt library from DigitalOcean that Dave suggested, I directly get the metadata once I have the name of the VM so no more iteration !

**Comment by aharivel** [comment]
Hi @dave-tucker and @sthaha ,
I've pushed a new version of my commit.
I ended up using digitalocean libvirt library and simplify the way I retrieve the metadata name.
I did not use any UUID. I think we (I?) can change that on a second phase.
My intention is to disturb the least possible the actual code while being able to evolve at the same time with the new requirement (UUID instead of name?)

Please tell me what you would like me to change on the base of that.

PS: I'll deal with the failing check as soon as they are finished.

**Comment by dave-tucker** [comment]
> My intention is to disturb the least possible the actual code while being able to evolve at the same time with the new requirement (UUID instead of name?)


Just confirming that ☝️ works for me.
Name from metadata using the digital ocean go library as step 1
We can open an issue to discuss using UUID as a future enhancement

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [reply pkg/libvirt/resolve_vm.go:117]: Sorry, I wasn't clear :( 
What I meant is that as a good practice we should keep variables close to its usage. E.g. `token` is initialised here but used may be 20 lines below. For a reader / maintainer look at the code, they have to keep a context of this when they read its usage 20 lines later.  But if the variables are moved closer to its actual usage, it reduces the cognitive load.

Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:117]: Oh ok I see. I'm a long year C developer some habits are hard to get rid of !
I'm moving the variable close to its usage.

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:137]: can't we `defer l.Disconnect()` instead?

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:141]:
```suggestion
		return "", fmt.Errorf("libvirt Domainlookup fail %w", err)
```

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:150]:
```suggestion
		return "", fmt.Errorf("failed to read metadata XML: %w", err)
```

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:158]:
```suggestion
	if nameElement == nil {
		return "", fmt.Errorf("no <%s> element found", token)
	} 
	return nameElement.Text(), nil

```

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:130]:
```suggestion
		return "", fmt.Errorf("libvirt failed to connect: %w", err)
```

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [inline pkg/libvirt/resolve_vm.go:125]: wouldn't this crash if len(parts) == 3 ?

**Review by sthaha — CHANGES_REQUESTED** [review]
lgtm overall! some minor comments to address.

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [reply pkg/libvirt/resolve_vm.go:117]: Ahem ahem :)

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:137]: I did that but the linter says I don't check the return value of "l.Disconnect()"...

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:125]:
```
vmids := []string{
		"machine-qemu-1234-name",
		"machine---5678-anothername",
		"machine-qemu-91011-differentname",
		"invalid-vmid",
		"machine-0000-lastname",
		"qemu-machine-",
		"qemu-machine--123456-last-",
	}

	for _, vmid := range vmids {
		parts := strings.Split(vmid, "-")
		if len(parts) < 4 {
			fmt.Printf("The vmid provided does not match machine-qemu-number-name: %v\n", vmid)
			continue
		}

		domainName := strings.Join(parts[3:], "-")
		fmt.Println(domainName)
	}
```
Output:
```
name
5678-anothername
differentname
The vmid provided does not match machine-qemu-number-name: invalid-vmid
The vmid provided does not match machine-qemu-number-name: machine-0000-lastname
The vmid provided does not match machine-qemu-number-name: qemu-machine-
123456-last-
```

Any suggestion of name that could be problematic ?

Comment (sthaha) [reply pkg/libvirt/resolve_vm.go:124]: @aharivel the snipped you pasted, has len(parts) < 4 instead of 3 (as in the code).

File: pkg/libvirt/resolve_vm.go
Comment (sthaha) [reply pkg/libvirt/resolve_vm.go:136]: we can silence the linter :) when we purposely ignore errors

Comment (sthaha) [reply pkg/libvirt/resolve_vm.go:136]: On second thought, how we return the error if Disconnect fails?

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:124]: I literally have no idea what happen there but thanks for pointing that out.

File: pkg/libvirt/resolve_vm.go
Comment (aharivel) [reply pkg/libvirt/resolve_vm.go:136]: I'm going to put the `defer l.Disconnect()` back into the code but the linter will failed on that part.

Comment (sthaha) [reply pkg/libvirt/resolve_vm.go]: @aharivel , I think we should return the error when running disconnect. A common pattern is to have something like 

```
func foo() (val, err error) {
   defer func() {
       err = l.Disconnect()
   }()
}
```

This way the caller gets to handle and report the error.

Comment (aharivel) [reply pkg/libvirt/resolve_vm.go]: Right, I do prefer this me too.

**Review by sthaha — COMMENTED** [review]
LGTM! Thanks a lot @aharivel !
I will let @dave-tucker take another look before merging.

**Comment by sthaha** [comment]
@aharivel in case you didn't notice, this PR requires rebasing

**Comment by aharivel** [comment]
> @aharivel in case you didn't notice, this PR requires rebasing

oh sorry, I thought I had no right to do that. Let me do it now. thanks !

**Comment by aharivel** [comment]
@sthaha I think the error above is not related to my patch. Am I wrong ?

**Review by sthaha — APPROVED** [review]
LGTM. I will let @dave-tucker do a final review before merging.

**Review by marceloamaral — APPROVED** [review]
Looks good to me too.
/lgtm

**Comment by aharivel** [comment]
It still needs you review @dave-tucker ! 
Everything should be fairly OK by now as already 2 out-of 3 reviewers approved the changes !

**Review by dave-tucker — APPROVED** [review]

**Comment by sthaha** [comment]
@aharivel , I was going to merge it this one but can't due to merge conflicts. 
Could you please rebase one last time (without any code changes 🙏 ) .

**Comment by aharivel** [comment]
Hey @sthaha , everything is green now :)

**Pull request merged**
